### PR TITLE
Feat/list objects shadow mode

### PR DESCRIPTION
**Pull request by cikasfm**
# Description

Enable experimental list-objects optimizations in shadow mode.

## Summary

This PR introduces **shadow mode** support for the ListObjects query in OpenFGA. Shadow mode allows both the standard and optimized ListObjects query implementations to be executed in parallel, with their results compared and logged for analysis. This feature is intended to help validate the correctness and performance of the optimized implementation before fully enabling it in production.

## Key Changes

- Added `shadowedListObjectsQuery`, which wraps both standard and optimized ListObjects queries.
- Introduced configuration options for enabling shadow mode, sampling percentage, and timeout.
- Both standard and optimized queries are executed in parallel when shadow mode is enabled; results and latencies are logged and compared.
- No changes to the public API; shadow mode is opt-in via configuration.
- Added and updated unit tests to cover shadow mode scenarios and edge cases.
- Improved documentation for new and existing code related to ListObjects queries.

## Motivation

Shadow mode provides a safe way to test and validate the optimized ListObjects query implementation in real-world scenarios, without impacting end users. It helps identify discrepancies and performance regressions before rolling out optimizations broadly.

## Testing

- Unit tests cover:
  - Shadow mode enabled/disabled
  - Error handling for both standard and optimized queries
  - Configuration edge cases

## Usage

To enable this feature, use the following server config:

```
	listObjectsOptions := []openfgaserver.OpenFGAServiceV1Option{
		openfgaserver.WithShadowListObjectsQueryEnabled(true),
		openfgaserver.WithShadowListObjectsQueryTimeout(10*time.Second),
		openfgaserver.WithShadowListObjectsQuerySamplePercentage(10),
	}
	openFGAServerOptions = append(openFGAServerOptions, listObjectsOptions...)
```

## Related Issues

- Enabling Weighted Graph use

## Checklist

- [x] Code compiles and passes all tests
- [x] Documentation updated
- [x] Unit tests added/updated

---

**Reviewer notes:**  
Please review the shadow mode logic, configuration handling, and test coverage. Feedback on logging and sampling approach is welcome.

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

## Summary by CodeRabbit

* **New Features**
  * Introduced experimental shadow mode for list object queries, enabling concurrent comparison of standard and optimized query implementations without impacting production results.
  * Added configuration options for shadow mode, including enablement flag, sampling percentage, timeout, and maximum delta items for result comparison.
  * Enabled experimental list object optimizations via the `enable-list-objects-optimizations` flag.

* **Bug Fixes**
  * Enhanced validation of shadow mode configuration parameters to enforce valid ranges and prevent misconfiguration.

* **Documentation**
  * Updated changelog to document the addition of experimental list object optimizations in shadow mode.

* **Tests**
  * Added extensive tests covering shadow mode functionality, configuration validation, concurrency handling, and result comparison logic.

### CODE REVIEW

File: pkg/server/commands/list_objects.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ var (
 	})
 )
 
-type ListObjectsQuery struct {
+type listObjectsQuery struct {
```
Comment (senojj) [inline pkg/server/commands/list_objects.go]: Why not export this type?

File: pkg/server/commands/list_objects.go [hunk 2]
```diff
@@ -95,72 +108,86 @@ type ListObjectsResponse struct {
...
+		d.listObjectsOptimizationEnabled = enabled
+	}
+}
+
 func NewListObjectsQuery(
 	ds storage.RelationshipTupleReader,
 	checkResolver graph.CheckResolver,
 	opts ...ListObjectsQueryOption,
-) (*ListObjectsQuery, error) {
+) (ListObjectsQuery, error) {
```
Comment (senojj) [inline pkg/server/commands/list_objects.go]: Why return an interface instead of the concrete type?

File: pkg/server/commands/list_objects.go (hunk 1 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: @senojj the goal is to "relieve" the client of need to know the internals of the object returned, whether it's a `ListObjectQuery` or `ShadowedListObjectsQuery` and just work with the interface.

The reason for that is that based on config, a different object type can be returned in `NewListObjectsQueryWithShadowConfig` and for consistency I'm returning the same in `NewListObjectsQuery`.

File: pkg/server/commands/list_objects.go (hunk 2 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: @senojj the goal is to "relieve" the client of need to know the internals of the object returned, whether it's a `ListObjectQuery` or `ShadowedListObjectsQuery` and just work with the interface.

The reason for that is that based on config, a different object type can be returned in `NewListObjectsQueryWithShadowConfig` and for consistency I'm returning the same in `NewListObjectsQuery`.

File: pkg/server/commands/list_objects.go (hunk 1 above)
Comment (senojj) [reply pkg/server/commands/list_objects.go]: The `listObjectsQuery` type has unexported fields, so the user cannot know the internals of the type, even when the type itself is exported.

`NewListObjectsQuery` only ever returns a `*listObjectsQuery` so there is no benefit to returning an interface from it. By returning an interface from this function, you prematurely constrain how the value may be useful to the user.

However, if you have function that can accept as an argument or return multiple types, polymorphism makes total sense.

File: pkg/server/commands/list_objects.go (hunk 2 above)
Comment (senojj) [reply pkg/server/commands/list_objects.go]: The key thing to note here is that this function will not ever return more than one type. Polymorphism serves no purpose here.

It makes sense for `NewListObjectsQueryWithShadowConfig`, but not here. Applying consistency only for the sake of consistency has no notable value.

File: pkg/server/commands/list_objects_shadow.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	return func(c *ShadowListObjectsQueryConfig) {
+		c.enabled = enabled
+	}
+}
+
+// WithShadowListObjectsQuerySamplePercentage sets the percentage of list_objects requests that will trigger the shadow query.
+func WithShadowListObjectsQuerySamplePercentage(samplePercentage int) ShadowListObjectsQueryOption {
+	return func(c *ShadowListObjectsQueryConfig) {
+		if samplePercentage < 0 || samplePercentage > 100 {
+			log.Fatalf("Invalid value for sample percentage value=%d, must be between 0 and 100", samplePercentage)
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: instead of crash and burn here, should we move the validation to when config is specified and assume that it is valid by the time we hit here?

File: pkg/server/commands/list_objects_shadow.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		defer wg.Done()
+		defer func() {
+			if r := recover(); r != nil {
+				err2 = fmt.Errorf("panic in fn2: %v", r)
+			}
+		}()
+		result2, err2 = fn2()
+		latency2 = time.Since(start)
+	}()
+	wg.Wait()
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: Instead of waiting for both the regular and the shadowed result to complete, would it make sense to wait for the regular to complete and then run the shadow in go func?  See https://github.com/openfga/openfga/blob/main/internal/graph/shadow_resolver.go

File: pkg/server/commands/list_objects_shadow.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+	if errOptimized != nil {
+		q.logger.Error("shadowed list objects error",
+			zap.String("func", fnName),
+			zap.Error(errOptimized))
+		return result, nil
+	}
+
+	q.logger.Info("shadowed list objects result",
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: Instead of always logging the shadow result, should we only flag when there are deltas?  See https://github.com/openfga/openfga/blob/90fdaeb6b5a53c7a4089cfac4b01ccd60e738c82/internal/graph/shadow_resolver.go#L97

Comment (adriantam) [reply pkg/server/commands/list_objects_shadow.go]: With that said, we also want to highlight if the number of items hit max limit / latency  / latencyOptimized is near the list objects timeout because those case will likely produce non-identical result

File: pkg/server/commands/list_objects_shadow.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+		return q.standard.Execute(ctx, req)
+	}
+	var fnOptimized = func(ctx context.Context) (*ListObjectsResponse, error) {
+		return q.optimized.Execute(ctx, req)
+	}
+
+	return executeShadowMode(ctx, q, "Execute", fnStandard, fnOptimized)
+}
+
+func (q *shadowedListObjectsQuery) ExecuteStreamed(ctx context.Context, req *openfgav1.StreamedListObjectsRequest, srv openfgav1.OpenFGAService_StreamedListObjectsServer) (*ListObjectsResolutionMetadata, error) {
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: ExecuteStreamed probably should not run in shadow mode because the result will inherently be different because list objects item will not return in predictable order

File: pkg/server/commands/list_objects_shadow.go (hunk 4 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go]: We're only waiting for the standard to complete, then cancel context for the shadow one - so it should stop running any DB queries & return immediatelly.

File: pkg/server/commands/list_objects_shadow.go (hunk 3 above)
Comment (justincoh) [reply pkg/server/commands/list_objects_shadow.go]: +1 to this, you could add something [here](https://github.com/openfga/openfga/blob/main/pkg/server/config/config.go#L400)

File: pkg/server/commands/list_objects_shadow.go (hunk 4 above)
Comment (justincoh) [reply pkg/server/commands/list_objects_shadow.go]: +1 to Adrian's point. If standard ListObjects errors, hits the timeout, or returns `maxResults`, then there's no value in running the shadow version for comparison, so it would be wasted effort even if we bail out early.

File: pkg/server/commands/list_objects_shadow.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+		zap.Duration("latencyOptimized", latencyOptimized))
+
+	return result, nil
+}
+
+// runInParallel helper to run two functions in parallel and collect their results and latencies.
+func runInParallel[T any](
+	fn1 func() (T, error),
+	fn2 func() (T, error),
+) (latency1, latency2 time.Duration, result1, result2 T, err1, err2 error) {
```
Comment (justincoh) [inline pkg/server/commands/list_objects_shadow.go]: That's a lot of return values. Can we do the comparisons in here?

File: pkg/server/commands/list_objects_shadow.go (hunk 4 above)
Comment (elbuo8) [reply pkg/server/commands/list_objects_shadow.go]: I'm confused, why would the successful "standard" cancel the shadow? It's not meant to be a race, and you def want to compute their result sets 🤔

File: pkg/server/commands/list_objects_shadow.go (hunk 7 above)
Comment (elbuo8) [reply pkg/server/commands/list_objects_shadow.go]: Can't say I disagree with this 🤔

File: pkg/server/commands/list_objects_shadow.go (hunk 4 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go]: IMHO we're hoping for the optimized to be faster, so we're not waiting for a slower result and ignoring it.

@adriantam do you think we should run "run & forget" the shadow impl once the current is completed? My thought was to run both in parallel & then compare the results. Let's discuss

**Comment by cikasfm** [comment]
@coderabbitai full review

File: pkg/server/commands/list_objects.go (hunk 1 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: https://martinfowler.com/bliki/TwoHardThings.html

File: pkg/server/commands/list_objects.go (hunk 2 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: https://martinfowler.com/bliki/TwoHardThings.html

File: pkg/server/commands/list_objects.go [hunk 8]
```diff
@@ -94,6 +107,19 @@ type ListObjectsResponse struct {
 	ResolutionMetadata ListObjectsResolutionMetadata
 }
 
+func (r *ListObjectsResponse) Clone() *ListObjectsResponse {
+	// Create a new instance of ListObjectsResponse
+	clone := &ListObjectsResponse{
+		Objects:            make([]string, len(r.Objects)),
+		ResolutionMetadata: *(NewListObjectsResolutionMetadata()),
```
Comment (senojj) [inline pkg/server/commands/list_objects.go]: I know this isn't part of your code changes. But does it really make sense for `NewListObjectsResolutionMetadata` to return a pointer? `ListObjectsResolutionMetadata` is a struct containing 3 pointers, which is pretty small, and there is nothing that would prevent it from being copied properly.

Here, the `ListObjectsResolutionMetadata` value is escaping to the heap in the call to `NewListObjectsResolutionMetadata()`, and then it is being dereferenced immediately, so it is copied, and added to a `ListObjectsResponse` instance, which escapes to the heap in `Clone()`.... In reality, none of these values need to be pointers and can exist on the stack perfectly well.

Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: makes sense

File: pkg/server/commands/list_objects.go [hunk 9]
```diff
@@ -155,10 +181,24 @@ func WithListObjectsDatastoreThrottler(threshold int, duration time.Duration) Li
...
+		d.listObjectsOptimizationEnabled = enabled
+	}
+}
+
 func NewListObjectsQuery(
 	ds storage.RelationshipTupleReader,
 	checkResolver graph.CheckResolver,
 	opts ...ListObjectsQueryOption,
+) (*ListObjectsQuery, error) {
+	return newListObjectsQuery(ds, checkResolver, opts...)
```
Comment (adriantam) [inline pkg/server/commands/list_objects.go]: curious why do we need to have `newListObjectsQuery`?

File: pkg/server/commands/list_objects_shadow.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	checkResolver graph.CheckResolver,
+	shadowConfig *ShadowListObjectsQueryConfig,
+	opts ...ListObjectsQueryOption,
+) (ListObjectsResolver, error) {
+	if shadowConfig == nil {
+		return nil, errors.New("shadowConfig must be set")
+	}
+	standard, err := newListObjectsQuery(ds, checkResolver,
+		// force disable optimizations
+		slices.Concat(opts, []ListObjectsQueryOption{WithListObjectsOptimizationEnabled(false)})...,
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go:117]: I know what you are trying to do...but a bit weird to force optimization to be false.  The concern is that if someone enabled shadow mode but set the percent to 0 (thinking this is the same thing as disable the shadow mode), we will force the optimization to be off.

File: pkg/server/commands/list_objects_shadow.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	// so it is impossible to compare the results
+	if loq, ok := q.main.(*ListObjectsQuery); ok {
+		if len(res.Objects) == int(loq.listObjectsMaxResults) {
+			q.logger.DebugWithContext(parentCtx, "shadowed list objects query skipped due to max results reached",
+				// common args
+				commonFields...,
+			)
+			return
+		}
+	}
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go:214]: we also want to skip if the initial query takes close to listObjectsDeadline as this indicates we will not get the complete result.

File: pkg/server/commands/list_objects.go [hunk 12]
```diff
@@ -94,6 +107,19 @@ type ListObjectsResponse struct {
 	ResolutionMetadata ListObjectsResolutionMetadata
 }
 
+func (r *ListObjectsResponse) Clone() *ListObjectsResponse {
```
Comment (senojj) [inline pkg/server/commands/list_objects.go]: Why does the `Clone` method need to return a pointer?

File: pkg/server/commands/list_objects_shadow.go [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+	if err != nil {
+		return obj, err
+	}
+	var cloned T
+	err = json.Unmarshal(marshalled, &cloned) // Unmarshal back to the original type.
+	if err != nil {
+		return obj, err
+	}
+	return cloned, nil
+}
```
[the remark this answers is not in the document]
Comment (senojj) [reply pkg/server/commands/list_objects_shadow.go]: I agree with this comment

File: pkg/server/commands/list_objects_shadow.go [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+		go func() {
+			q.executeShadowModeAndCompareResults(cloneCtx, cloneReq, cloneRes)
+			q.wg.Done() // only used for testing signals
+		}()
+	}
+	return res, err
+}
+
+// cloneObj is a utility function that clones an object of type T using JSON marshalling and unmarshalling.
+func cloneObj[T any](obj T) (T, error) {
```
Comment (senojj) [inline pkg/server/commands/list_objects_shadow.go]: Why are you cloning via JSON serialization? This is extremely inefficient.

File: pkg/server/commands/list_objects_shadow.go [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+		zap.String("store_id", req.GetStoreId()),
+		zap.String("model_id", req.GetAuthorizationModelId()),
+	}
+	defer func() {
+		if r := recover(); r != nil {
+			q.logger.ErrorWithContext(parentCtx, "panic recovered",
+				append(commonFields, zap.Any("error", r))...,
+			)
+		}
+	}()
```
Comment (senojj) [inline pkg/server/commands/list_objects_shadow.go:201]: 1. Is it ok for the application to continue after a panic?
2. `recover` is an expensive operation. Do you expect a panic somewhere in this function?

File: pkg/server/commands/list_objects_shadow.go [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+
+	if errShadow != nil {
+		q.logger.WarnWithContext(parentCtx, "shadowed list objects error",
+			append(commonFields, zap.Any("error", errShadow))...,
+		)
+		return
+	}
+
+	objects := copyAndSortArr(res.Objects)
+	shadowObj := []string{}
```
Comment (senojj) [inline pkg/server/commands/list_objects_shadow.go]: You can avoid an allocation here by initializing `shadowObj` to a `nil` value.

File: pkg/server/commands/list_objects_shadow.go [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+	if shadowRes != nil {
+		shadowObj = copyAndSortArr(shadowRes.Objects)
+	}
+	// compare sorted string arrays - sufficient for equality check
+	if !slices.Equal(objects, shadowObj) {
```
Comment (senojj) [inline pkg/server/commands/list_objects_shadow.go:247]: is the point of this to test that `objects` was sorted to begin with?

File: pkg/server/commands/list_objects_shadow.go [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+	}
+}
+
+// calculateDelta compares two slices of strings and returns a slice of strings representing the delta.
+// Each string in the delta is prefixed with "+" if it exists in the shadow but not in the main,
+// or "-" if it exists in the main but not in the shadow.
+// Used for logging differences between the main and shadow results in the shadowed list objects query.
+func calculateDelta(inputMain []string, inputShadow []string) []string {
+	mainObjects := make(map[string]bool)
+	shadowObjects := make(map[string]bool)
```
Comment (senojj) [inline pkg/server/commands/list_objects_shadow.go]: Not critical, but you can make these maps more space efficient by making them `map[string]struct{}`

File: pkg/server/commands/list_objects.go (hunk 12 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: Mostly because the pointer is used as argument in all places

File: pkg/server/commands/list_objects.go (hunk 9 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects.go]: no need any more after 10x refactoring...

File: pkg/server/commands/list_objects_shadow.go (hunk 14 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go]: I agree! Maybe we don't need to clone the request at all. Otherwise, the stack of code required just to clone would be more than I think it's valuable, just for shadow mode

File: pkg/server/commands/list_objects_shadow.go (hunk 15 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go:201]: obviously it's only for unexpected panics. The goal here is not to kill the app just because something failed in shadow mode.

File: pkg/server/commands/list_objects_shadow.go (hunk 11 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go:214]: @adriantam I agree, but we need to define what is `initial query takes close to listObjectsDeadline`. Is it within 5%? Or within 5ms? Or do we want to make it configurable? Or does it even matter? At this point if the original query was very slow then we're probably timing out & failing with Context Deadline Exceeded, in which case we would not even run the shadow mode. For successful but slow - IMHO we want to see the performance gains. Do we really want to check if it "almost exceeded deadline"?

File: pkg/server/commands/list_objects.go (hunk 12 above)
Comment (senojj) [reply pkg/server/commands/list_objects.go]: By doing that, you are forcing the cloned value to escape to the heap. If you were to return a value and reference the value when passing it into a function, you could avoid a heap allocation in scenarios that don't require the lifetime of the value to extend beyond the lifetime of the current stack frame.

I realize that pointers are currently used prolifically throughout this project, but we should be more mindful about when we force heap allocations.

File: pkg/server/commands/list_objects_shadow.go (hunk 10 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go:117]: Good point! But also...that sounds like a valid use case to me. Is there a point to have "main" run optimizations enabled AND a shadow mode, with the same optimization enabled?

File: pkg/server/commands/list_objects_shadow.go (hunk 11 above)
Comment (adriantam) [reply pkg/server/commands/list_objects_shadow.go:214]: As discussed offline, we can make it within 100ms.  The reason we want to do this is that list objects will not return erorr, but instead incomplete response.  This can be done as a follow up PR

File: .golangci.yaml [hunk 19]
```diff
@@ -53,10 +53,6 @@ linters:
     staticcheck:
       checks:
         - all
-    revive:
```
Comment (adriantam) [inline .golangci.yaml]: why?

Comment (cikasfm) [reply .golangci.yaml]: oh, wait. I don't think this was intentional

File: pkg/server/commands/list_objects_shadow.go [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+	shadowEnabled bool          // A boolean flag to globally enable or disable the shadow mode for list_objects queries. When false, the shadow query will not be executed.
+	shadowPct     int           // An integer representing the shadowPct of list_objects requests that will also trigger the shadow query. This allows for controlled rollout and data collection without impacting all requests. Value should be between 0 and 100.
+	shadowTimeout time.Duration // A time.Duration specifying the maximum amount of time to wait for the shadow list_objects query to complete. If the shadow query exceeds this shadowTimeout, it will be cancelled, and its result will be ignored, but the shadowTimeout event will be logged.
+	maxDeltaItems int           // The maximum number of items to log in the delta between the main and shadow results. This prevents excessive logging in case of large differences.
+	logger        logger.Logger
+}
+
+func NewShadowListObjectsQueryConfig(opts ...ShadowListObjectsQueryOption) *ShadowListObjectsQueryConfig {
+	result := &ShadowListObjectsQueryConfig{
+		shadowEnabled: false,                  // Disabled by default
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go:79]: nitpik (non-blocking): ideally this will share the default config instead of hard code value

File: pkg/server/commands/list_objects_shadow.go [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+						zap.Any("request", req),
+						zap.String("store_id", req.GetStoreId()),
+						zap.String("model_id", req.GetAuthorizationModelId()),
+						zap.Any("error", r),
+					)
+				}
+			}()
+			defer q.wg.Done() // only used for testing signals
+
+			cloneRes := res.Clone()
```
[the remark this answers is not in the document]
Comment (senojj) [reply pkg/server/commands/list_objects_shadow.go]: @cikasfm did you intend to zero-out the resolution metadata on clone?

File: pkg/server/commands/list_objects_shadow.go [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+		return nil, err
+	}
+
+	result := &shadowedListObjectsQuery{
+		main:          standard,
+		shadow:        optimized,
+		shadowPct:     shadowConfig.shadowPct,
+		shadowTimeout: shadowConfig.shadowTimeout,
+		logger:        shadowConfig.logger,
+		random:        rand.New(rand.NewSource(time.Now().UnixNano())),
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: why do we need to do this?  I thought that we don't need to seed anymore - https://pkg.go.dev/math/rand#Seed.
See https://go.dev/blog/randv2

File: pkg/server/commands/list_objects_shadow.go [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	// compare sorted string arrays - sufficient for equality check
+	if !slices.Equal(objects, shadowObj) {
+		delta := calculateDelta(objects, shadowObj)
+		if len(delta) > q.maxDeltaItems {
+			delta = delta[:q.maxDeltaItems] // limit the delta to maxDeltaItems
+		}
+		// log the differences if the shadow query failed or if the results are not equal
+		q.logger.InfoWithContext(parentCtx, "shadowed list objects result difference",
+			append(commonFields,
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go:245]: probably good to log the number of delta items

File: pkg/server/commands/list_objects_shadow.go [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+	shadowRes, errShadow := q.shadow.Execute(shadowCtx, req)
+
+	if errShadow != nil {
+		q.logger.WarnWithContext(parentCtx, "shadowed list objects error",
+			append(commonFields, zap.Any("error", errShadow))...,
+		)
+		return
+	}
+
+	objects := copyAndSortArr(mainResult)
```
Comment (adriantam) [inline pkg/server/commands/list_objects_shadow.go]: Since we are using the map to calculate the delta anyways, would it make sense to just use the map?  (I.e., no need to do duplicate work of sorting)

File: pkg/server/commands/list_objects_shadow.go (hunk 22 above)
Comment (cikasfm) [reply pkg/server/commands/list_objects_shadow.go]: I think this was suggested by `@coderabbitai`; I guess we really don't

**Review by adriantam — APPROVED** [review]
